rse-py-scripting.Rmd

```diff
@@ -268,10 +717,210 @@ ERROR_MESSAGES[user_language]['cannot_read_file']
+Thus,
+if we run the small program shown below,
+only the warning message appears:
+
+```python
+import logging
+
+logging.warning('This is a warning.')
+logging.info('This is just for information.')
+```
+
+```text
+WARNING:root:This is a warning.
+```
+
+We can configure logging to send messages to a file instead of to standard error
+using `logging.basicConfig`.
+(This has to be done before we make any logging calls:
+it's not retroactive.)
+We can use the same function to set the logging level:
+everything at or above the specified level is displayed.
+
+```python
+import logging
+
+logging.basicConfig(level=logging.DEBUG, filename='everything.log')
+
+logging.debug('This is for debugging.')
+logging.info('This is just for information.')
+logging.warning('This is a warning.')
+logging.error('Something went wrong.')
+logging.critical('Something went seriously wrong.')
+```
+
+```text
+DEBUG:root:This is for debugging.
+INFO:root:This is just for information.
+WARNING:root:This is a warning.
+ERROR:root:Something went wrong.
+CRITICAL:root:Something went seriously wrong.
+```
+
+By default,
+`basicConfig` re-opens the log file in [append mode][append-mode];
+we can use `filemode='w'` to overwrite the existing log data.
+This is useful during debugging,
+but we should think twice before doing in production,
+since the information we throw away always turns out to have been exactly what we needed to find a bug.
+
+Many programs allow users to specify logging levels and log file names as command-line parameters.
+At its simplest,
+this is a single flag `-v` or `--verbose` that changes the logging level from `WARNING` (the default)
+to `DEBUG` (the noisiest level).
+There may also be a corresponding flag `-q` or `--quiet` that changes the level to `ERROR`,
+and a flag `-l` or `--logfile` that specifies a log file name
+(with standard error being the default for message output).
+
+> **`tail -f`**
+>
+> A handy trick during development is to configure logging to send messages to a file,
+> and then open another terminal window and run <code>tail -f <em>filename</em></code>
+> to display changes to that file as information is appended.
+> Doing this gives you a record of the log output
+> while allowing you to monitor your program's progress interactively.
+
+Libraries like `logging` can send messages to many destinations:
+[rotating files][rotating-file] so that the system always has messages from the last few hours
+but doesn't fill up the disk,
+for example,
+or a centralized logging server of some kind that collates logs from many different systems.
+We don't need any of these when we start,
+but the data engineers and system administrators who eventually have to install and maintain your programs
+will be very grateful that you used `logging`,
+because then they can set it up the way they want with very little work.
+
+## How should I log my program's configuration? {#rse-py-scripting-log-config}
+
```

Comment:
How does this section relate to the previous section: "How can I keep a record of the actual configuration that produced particular results? {#s:configure-dump}"

Reply:
Ah - good catch. Section replaced with a short callout box. Sorry about that.